internal/auth/auth.go

```diff
@@ -0,0 +1,172 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+
+	"github.com/pkg/browser"
+	"github.com/speakeasy-api/speakeasy/internal/config"
+)
+
+const (
+	// Replace these with a local server for development
+	appURL = "https://app.speakeasyapi.dev"
+	apiURL = "https://api.prod.speakeasyapi.dev"
+)
+
+type authResult struct {
+	config.SpeakeasyAuthInfo
+	err error
+}
+
+func Authenticate(force bool) error {
+	ctx := context.Background()
+
+	existingApiKey, setByEnvVar := config.GetSpeakeasyAPIKey()
+	if existingApiKey != "" && !force {
+		if err := testAuth(existingApiKey); err != nil {
+			if setByEnvVar {
+				return err
+			}
+		} else {
+			return nil
+		}
+	}
+
+	if !force {
+		fmt.Println("Authentication needed")
+	}
+
+	done := make(chan authResult)
+
+	addr, srv := startServer(done)
+	defer func() { _ = srv.Shutdown(ctx) }()
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+
+	url := fmt.Sprintf("%s?cli_callback_url=%s&cli_host_name=%s\n", appURL, addr, hostname)
+
+	if err := browser.OpenURL(url); err != nil {
+		fmt.Println("Please open the following URL in your browser:", url)
+	}
+
+	res := <-done
+	if res.err != nil {
+		return res.err
+	}
+
+	if err := testAuth(res.APIKey); err != nil {
+		return err
+	}
+
+	if err := config.SetSpeakeasyAuthInfo(res.SpeakeasyAuthInfo); err != nil {
+		return fmt.Errorf("failed to save API key: %w", err)
+	}
+
+	fmt.Println("Authentication successful!")
+
+	return nil
+}
+
+func Logout() error {
+	if err := config.ClearSpeakeasyAuthInfo(); err != nil {
+		return fmt.Errorf("failed to remove API key: %w", err)
+	}
+
+	fmt.Println("Logout successful!")
+
+	return nil
+}
+
+func testAuth(apiKey string) error {
+	// TODO eventually replace with a call from the SDK
+	u, err := url.JoinPath(apiURL, "v1/auth/validate")
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodGet, u, nil)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("x-api-key", apiKey)
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to successfully contact the Speakeasy Platform: %w", err)
```

Comment:
Contact speakeasy support ?

Reply:
@simplesagar add a suggestion if you don't mind of what you would like it to say

Reply:
have made a change let me know if that works for you